File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    d_r = [-1,0,1,0]
    d_c = [0,1,0,-1]
    q = deque([(0,0)])
    visited = [[0]*len(maps[0]) for _ in range(len(maps))]
    visited[0][0] = 1
    ans = -1
    while q:
        cur_r, cur_c = q.popleft()
        if cur_r == len(maps)-1 and cur_c == len(maps[0])-1 and visited[cur_r][cur_c]!= 0:
            ans = visited[cur_r][cur_c]
            break
        for d in range(4):
            n_r = cur_r + d_r[d]
            n_c = cur_c + d_c[d]
            if n_r in range(len(maps)) and n_c in range(len(maps[0])):
                if maps[n_r][n_c] == 1 and visited[n_r][n_c] == 0:
                    q.append((n_r,n_c))
                    visited[n_r][n_c] = visited[cur_r][cur_c] + 1
            continue
    return ans

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static int[][] map;
    static int[] d_r = {-1,0,1,0};
    static int[] d_c = {0,1,0,-1};
    static int[][] visited;
    public int solution(int[][] maps) {
        visited = new int[maps.length][maps[0].length];
        map = maps;
        return bfs(maps.length,maps[0].length);
    }
    public int bfs(int n,int m){
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0,0});
        visited[0][0] = 1;
        
        while(q.size()!=0){
            int[] s = q.poll();
            int cur_r = s[0];
            int cur_c = s[1];
            if(cur_r==n-1&&cur_c==m-1){
                return visited[cur_r][cur_c];
            }
            if(visited[cur_r][cur_c]==0){
                continue;
            }
            for(int i=0;i<4;i++){
                int n_r = d_r[i]+cur_r;
                int n_c = d_c[i]+cur_c;
                if(0<=n_r && n_r<n && 0<=n_c && n_c<m && map[n_r][n_c] == 1 && visited[n_r][n_c]==0){
                    q.add(new int[]{n_r,n_c});
                    visited[n_r][n_c] = visited[cur_r][cur_c]+1;
                }
            }
            
        }
        return -1;
    }
}